David Generic Practice - Weapon Selector

HanBrush opened the pull request:
David Generic Practice

HanBrush committed:
David Generic Practice - Weapon Selector

David Generic Practice

goodmorningcody commented on Assets/David/GenericPractice/Scripts/Events.cs:
```diff
@@ -0,0 +1,41 @@
+
+    }
+
+    public struct SelectWeaponTypeEvent<T> : WeaponTypeEvent where T : IWeaponType
```
멋져요! 제네릭을 완벽히 이해하셨군요!

goodmorningcody commented on Assets/David/GenericPractice/Scripts/MonoSkillButton.cs:
```diff
@@ -0,0 +1,24 @@
+    {
+        [SerializeField] private Text txt_skillName;
+
+        public void ButtonInit(string skill)
```
이런 코드도 onClick에 이벤트를 달아서 처리를 하지 않았다면 인자로 전달된 skill을 MonoSkillButton의 프로퍼티로 저장했어야 됐을꺼에요.

goodmorningcody commented on Assets/David/GenericPractice/Scripts/MonoWeaponButton.cs:
```diff
@@ -0,0 +1,25 @@
+        [SerializeField] Text txt_weaponName;
+        // Start is called before the first frame update
+
+        public void ButtonInit<T>(T weaponGadget) where T : IWeaponGadget
```
멋진 메소드네요@_@

goodmorningcody commented on Assets/David/GenericPractice/Scripts/WeaponGadgetModels.cs:
```diff
@@ -0,0 +1,63 @@
+    {
+        int Index { get; }
+        string ItemName { get; }
+    }
```
IWeaponGadget 인터페이스는 Index, ItemName 이라는 getter 프로퍼티의 구현을 강제하는 인터페이스인데요.
IWeaponGadget 인터페이스를 구현체 클래스의 생성자에서 this.Index/this.ItemName을 초기화하는 형태보단 생성자의 인자로 전달된 IWeapon을 프로퍼티를 통해 접근하는 방식으로도 구현할 수 있을것 같아요. 그렇다고해서 이 코드가 꼭 나쁘단 생각이 들진 않아요.

HanBrush replied:
아하 그러네요 굳이 Index 와 ItemName 두 개의 프로퍼티가 필요없이 IWeapon 프로퍼티 하나만 받아도 되겠군요!

goodmorningcody replied:
IWeapon 프로퍼티 하나로 받으면 다른 문제가 생기는데요. 외부에서 IWeapon 프로퍼티의 Index/ItemName에 접근하게 되는 문제에요. IWeapon 구현체인 Sword가 있다고 했을 때, Sword의 이름에 접근하게 위해 아래와 같은 코드가 생기는거죠.

Sword mySword = new Sword()
mySword.weapon.Index;
mySword.weapon.ItemName;

그래서 저는 IWeapon의 Interface를 Extension해서 mySword.GetIndex()와 같은 형태로 사용할 수 있도록 만드는 편이에요.
public static class IWeaponExtension
{
    public static int GetIndex(this IWeapon weapon) { return weapon.Index; }
    public static int GetItemName(this IWeapon weapon) { return weapon.ItemName; }
}

위와 같이 interface의 extension을 자주 쓰는 이유가 있는데요.
흔히 모던 랭귀지라고 부르는 언어들은 "interface가 default implementation을 갖을 수 있다."라는 스펙을 가지고 있어요.
예를 들어 해당 스펙은 interface를 필수 구현하지 않아도 되게 만들어서 interface의 제약을 약화시키는데요.
그렇게 되면 interface를 abstract class 처럼 쓸 수 있게 해줘요.
특히 C# 최신 스펙을 보면 interface가 default implementation을 갖을 수 있게 되면서 최신 스펙에서는 interface를 extension을 궂이 하지 않아도 됐어요. 그래서 interface의 extension이 C#의 객체지향을 파괴한다고 생각하지 않아요.
https://devblogs.microsoft.com/dotnet/default-implementations-in-interfaces/

goodmorningcody commented on Assets/David/GenericPractice/Scripts/WeaponGadgetModels.cs:
```diff
@@ -0,0 +1,63 @@
+
+        public string ItemName { get; private set; }
+
+        public WandGadget(Wand wand)
```
Wand 타입을 받는게 아니라 IWeapon 타입 인터페이스로 받아야 되는거죠?

HanBrush replied:
앗 넵 맞습니다 ㅎㅎ IWeapon 타입으로 받는겁니다

goodmorningcody commented on Assets/David/GenericPractice/Scripts/WeaponSelector.cs:
```diff
@@ -0,0 +1,151 @@
+        TypeSword typeSword;
+        TypeBow typeBow;
+        TypeAx typeAx;
+        TypeWand typeWand;
```
Type\*와 같은 네이밍은 적절하지 않은것 같아요.
흔히 \*Type 혹은 Type\*과 같은 형태의 네이밍은 enum을 이야기하는 경우가 많아서 그러합니다.

HanBrush replied:
코딩할 때 참고하도록 하겠습니다!

HanBrush committed:
cody 코드리뷰 반영 후 수정

IWeaponExtension 클래스 추가,
WandGadget 클래스 생성자 인자 타입 수정 (Wand -> IWeapon)
TypeSword, TypeBow, TypeAx, TypeWand 클래스 이름 변경 => WeaponSword, WeaponBow, WeaponAx, WeaponWand,
